File: Numeros Pseudo-Aleatorios/Middle Square/MiddleSquare.go
/*
 PRNG .......: Middle Square
 Ano ........: 1946
 Autor ......: John von Neumann
*/

package main

import (
	"fmt"
	"time"
)

func meanSquare(seed int64) int64 {
	seed = (seed * seed / 100) % 10000
	return (seed + 10000) % 10000
}

func main() {
	var Xn1 int64 = 1234
	var Xn2 int64 = time.Now().UnixNano() / int64(time.Millisecond)

	fmt.Println("GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)")
	fmt.Println("PSEUDORANDOM NUMBER GENERATOR (PRNG)")
	fmt.Println("PADRAO: Middle Square")
	fmt.Println("==========================================")
	fmt.Println()
	fmt.Println("    Semente: 1234        Semente: TIME")
	fmt.Println("    -------------        -------------")

	for i := 1; i <= 10; i++ {
		Xn1 = meanSquare(Xn1)
		Xn2 = meanSquare(Xn2)
		fmt.Printf("%17d        %13d\n", Xn1, Xn2)
	}
}

File: Numeros Pseudo-Aleatorios/Middle Square/MiddleSquare.pas
{ 
 PRNG .......: Middle Square
 Ano ........: 1946
 Autor ......: John von Neumann
}

program Mean_Square;

uses
  SysUtils, DateUtils;
  
function MeanSquare(var seed: LongInt): LongInt;
begin
  seed := (seed * seed div 100) mod 10000;
  MeanSquare := seed;
end;

var
  Xn1, Xn2: LongInt;
  i: Integer;

begin

  Xn1 := 1234; 
  Xn2 := LongWord(DateTimeToUnix(Now)); 

  writeln('GERADOR DE NUMEROS PSEUDO ALEATORIOS (GNA)');
  writeln('PSEUDORANDOM NUMBER GENERATOR (PRNG)');
  writeln('PADRAO: Middle Square');
  writeln('==========================================');
  writeln;
  writeln('    Semente: 1234        Semente: TIME');
  writeln('    -------------        -------------');

  for i := 1 to 10 do
    writeln(MeanSquare(Xn1):17, '        ', MeanSquare(Xn2):13);

end.
